src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -112,6 +104,168 @@ int IndexOf(string str, char c)
             }
         }
 
+        private static bool TryGetNormalizedVersion(string str, [NotNullWhen(true)] out Version? version)
+        {
+            if (string.IsNullOrWhiteSpace(str))
+            {
+                version = null;
+                return false;
+            }
+
+            int minor = 0;
+            int build = 0;
+            int revision = 0;
+
+            // Check for all the possible parts of the version string. If lastParsedPosition is less than the end of
+            // the string, the input string was invalid (e.g. "1.2.3.4.5").
+            bool success = ParseSection(str, 0, out int lastParsedPosition, out int major) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out minor) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out build) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out revision) &&
+                lastParsedPosition == str.Length;
+
+            if (success)
+            {
+                version = new Version(major, minor, build, revision);
+                return true;
+            }
+            else
+            {
+                version = null;
+                return false;
+            }
+
+            // Returns false if an invalid section was found while processing the string.
+            static bool ParseSection(string str, int start, out int end, out int versionNumber)
```

Comment:
I don't want to shift goalposts, wasting time & effort, so consider this more like a conversation than a request for change. But I discovered in another PR that System.Memory, and therefore `Span<char>` is available thoughout NuGet (the dll is already in our VS install, for our MSBuild task, and is ilmerged in nuget.exe).

I find this `ParseSection` method a bit long, so I wonder if using `Span<char>` could make the implementation either shorter, or split into smaller methods, making it easier to read & understand. Although, from a perf point of view, I imagine this forward only, read once implementation will be hard to beat.

Reply:
That's an interesting point.  I think I'll merge this for now and maybe later someone can experiment with spans to see if the code can be cleaner and/or more performant.